### DISCUSSION TITLE
How clickhouse process duplicates in ReplicatedMergeTree

### DISCUSSION THREAD

**Discussion by gaoxingliang** [ClickHouse/ClickHouse#53311] [Q&A] [answered]
I implemented the [ebay block aggregator](https://tech.ebayinc.com/engineering/block-aggregator-real-time-data-ingestion-from-kafka-to-clickhouse-with-deterministic-retries/) in java to transfer data from kafka to clickhouse.

The main idea to remove duplicates is:
> Specifically, ClickHouse deduplicates “identical blocks” (i.e., blocks with the same size), containing the same rows in the same order. If we somehow make sure that our retries are in the form of identical blocks, then we avoid data duplication while we are re-sending data. 

I also found the block information is recorded at zookeeper path ` /clickhouse/tables/0/default/test_log3/blocks`

So my question is:
1. Will the blocks information merged? when merge happened? If merged, we cant use this strategy to remove duplicates.
2. If the blocks information in zk is not deleted. Will this cause a problem if we have a lot of small inserts?  as time goes on, the nodes in the zk path increase.

**Comment by gaoxingliang** [comment]
the max num of children nodes in zk: https://stackoverflow.com/questions/29791134/zookeeper-max-number-of-children-per-node

**Reply by UnamedRus** [reply, marked as answer]
> Will the blocks information merged? when merge happened? If merged, we cant use this strategy to remove duplicates.

Block information only affect inserts of data. 

> If the blocks information in zk is not deleted. Will this cause a problem if we have a lot of small inserts? as time goes on, the nodes in the zk path increase.

Older blocks being removed after some time/ number for blocks

```
                                               │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
│ replicated_deduplication_window                           │ 100    │       0 │ How many last blocks of hashes should be kept in ZooKeeper (old blocks will be deleted).
                                               │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
│ replicated_deduplication_window_seconds                   │ 604800 │       0 │ Similar to "replicated_deduplication_window", but determines old blocks by their lifetime. Hash of an inserted block will be deleted (and the block will not be deduplicated after) if it outside of one "window". You can set very big replicated_deduplication_window to avoid duplicating INSERTs during that period of time. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
│ replicated_deduplication_window_for_async_inserts         │ 10000  │       0 │ How many last hash values of async_insert blocks should be kept in ZooKeeper (old blocks will be deleted).
                                               │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
│ replicated_deduplication_window_seconds_for_async_inserts │ 604800 │       0 │ Similar to "replicated_deduplication_window_for_async_inserts", but determines old blocks by their lifetime. Hash of an inserted block will be deleted (and the block will not be deduplicated after) if it outside of one "window". You can set very big replicated_deduplication_window to avoid duplicating INSERTs during that period of time. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64 │
└───────────────────────────────────────────────────────────┴────────┴─────────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴──────┴──────┴──────────┴────────┘

```

**Reply by gaoxingliang** [reply]
Thanks for your clarification.
